File: internal/console/ui/app/screens/staff.js
/* ==========================================================================
   Who can open this console.

   # THE QUESTION THIS CONSOLE COULD NOT ASK ABOUT ITSELF

   `cmd/staff list` has printed this roster since phase 0, into a terminal
   beside the database. So the console could set what every student pays, erase
   a person, quarantine a question and read every entry in the audit — and could
   not answer "who else can do all of that". An access list nobody can see is an
   access list nobody reviews, and the row that matters on it is always the one
   nobody remembered was there.

   # IT IS NOT WHAT K-22 REFUSES

   That decision is about producing a page of STUDENTS from something somebody
   typed, and its argument is that browsing personal data cannot be told from
   working. It does not reach here: this is the platform's own access-control
   list, and there is no version of reviewing one that goes an address at a time
   — the whole question is who is on it that nobody thought to ask about.

   # AND THERE IS NO FORM, WHICH THE SCREEN SAYS OUT LOUD

   Granting lives in `cmd/staff`, which has to exist regardless: the first owner
   cannot be granted a role by a console that needs a role to open. A second
   door onto the same table would be a second thing to keep audited and in
   agreement, bought for the convenience of not opening a terminal on the rarest
   write this platform has. The sentence comes from the server, so somebody
   looking for the button finds the reason rather than concluding it was
   forgotten.

   # THE COLUMN WORTH READING IS THE LAST ONE

   Role and enrolment say what somebody CAN do. Neither says whether they have.
   A role granted a year ago and never used since is access nobody is missing,
   and it is the row an access review exists to find — so it is drawn as a
   finding rather than as a date, and the row is marked.
   ========================================================================== */

import { esc } from '../dom.js';
import { get } from '../request.js';

export default async function staff(section) {
  const el = document.createElement('div');
  el.className = 'view';

  el.innerHTML =
    '<header class="view-head">' +
      '<span class="eyebrow mono">Govern</span>' +
      '<h1>Who can open this</h1>' +
      '<p>Everybody with a role on this platform, and everybody who has had ' +
      'one. A role opens nothing without a second factor, and neither says ' +
      'whether it has been used — which is the column an access review is ' +
      'actually for.</p>' +
    '</header>' +
    '<div id="body" aria-live="polite"><p class="checking">Reading…</p></div>';

  const body = el.querySelector('#body');

  let answer;
  try {
    answer = await get('/console/api/v1/staff');
  } catch (e) {
    body.innerHTML = '<section class="block"><p class="none">' + esc(e.message) + '</p></section>';
    return { title: section.name, el };
  }

  const all = answer.staff || [];

  /* AN EMPTY ROSTER IS A DEFECT AND NOT AN EMPTY STATE. This page was opened by
     somebody holding a live staff row, so "nobody has a role" cannot be true —
     and it is a sentence somebody would act on. The server says so in its own
     words, because it is a statement about the system rather than about a page. */
  if (all.length === 0) {
    body.innerHTML = '<section class="block"><p class="none">'
      + esc(answer.impossible || '') + '</p></section>';
    return { title: section.name, el };
  }

  body.innerHTML =
    '<section class="block">' +
      /* THE SCROLL CONTAINER IS FOCUSABLE HERE AND IS NOT ON THE OTHER TABLES,
         and the difference is what is INSIDE them rather than a preference.

         `table-wrap` scrolls sideways — it is the console's one element allowed
         to, so the page never does — and a region that scrolls has to be
         reachable by keyboard. Everywhere else in this console the rows contain
         links: the history links to an entry, the record links to a person, and
         tabbing through them scrolls the box as a side effect. This table is six
         columns of facts with nothing to click, so there is nothing for a
         keyboard to land on and the container has to take the focus itself.

         WHICH MEANS IT NEEDS A NAME. A focusable box announced as "region" and
         nothing else is a stop on the tab order that says nothing about why it
         is there. `a11y-test` found this — the table went out without it. */
      '<div class="table-wrap" tabindex="0" role="region" ' +
        'aria-label="Everybody with a role on this platform">' +
      '<table class="grid">' +
        '<thead><tr>' +
          '<th scope="col">Who</th>' +
          '<th scope="col">Role</th>' +
          '<th scope="col">Can get in</th>' +
          '<th scope="col">Since</th>' +
          '<th scope="col">Let in by</th>' +
          '<th scope="col">Last opened this</th>' +
        '</tr></thead>' +
        '<tbody>' + all.map(row).join('') + '</tbody>' +
      '</table></div>' +
      '<p class="list-count">' + all.length
        + (all.length === 1 ? ' row' : ' rows') + '</p>' +
    '</section>' +

    '<section class="block">' +
      '<div class="block-top"><h2>Reading this list</h2></div>' +
      '<p class="aside">' + esc(answer.about_reviewing || '') + '</p>' +
    '</section>' +

    '<section class="block">' +
      '<div class="block-top"><h2>Why there is no form</h2></div>' +
      '<p class="aside">' + esc(answer.how_to_change_it || '') + '</p>' +
    '</section>';

  return { title: section.name, el };
}

function row(one) {
  const gone = Boolean(one.revoked_at);

  return '<tr>' +
    /* THE NAME READS AND THE ADDRESS IDENTIFIES, which is this console's
       two-line cell everywhere it shows a person. The address is the half that
       matters here: it is what `staff grant` and `staff revoke` take, so
       somebody acting on this row is copying that line. */
    '<td>' +
      '<span class="cell-main">' + esc(one.name || '—') +
        (gone ? '<span class="tag tag-quiet">left</span>' : '') +
      '</span>' +
      '<span class="cell-sub mono">' + esc(one.email) + '</span>' +
    '</td>' +

    // THE ROLE AS THE SERVER SAID IT. There are three, and a fourth would be a
    // decision about a person — drawn as itself rather than folded into the
    // nearest of the three, so it appears as a thing nobody has styled yet.
    '<td class="mono">' + esc(one.role) + '</td>' +

    '<td>' + access(one, gone) + '</td>' +

    '<td>' + esc(day(one.granted_at)) +
      (gone ? '<span class="cell-sub">until ' + esc(day(one.revoked_at)) + '</span>' : '') +
    '</td>' +

    /* NOBODY, SPELLED OUT. `staff.granted_by` is null for exactly one row in
       any deployment's life — the first owner, granted from a terminal by
       somebody who is not in this system — and a blank cell there reads as a
       field that failed to load. */
    '<td>' +
      (one.granted_by_email
        ? '<span class="cell-main">' + esc(one.granted_by_name || '—') + '</span>' +
          '<span class="cell-sub mono">' + esc(one.granted_by_email) + '</span>'
        : '<span class="none">nobody — the first owner</span>') +
    '</td>' +

    '<td>' + used(one, gone) + '</td>' +
  '</tr>';
}

/* WHETHER THE ROLE ACTUALLY OPENS ANYTHING, which is not the same fact as
   having one. The check is at the door: an account with a role and no second
   factor cannot reach a staff route, and there is no state in between. A screen
   that showed the role alone would describe access that does not exist. */
function access(one, gone) {
  if (gone) return '<span class="none">revoked</span>';
  if (!one.second_factor) {
    return '<span class="tag tag-warn">no second factor</span>';
  }
  return '<span class="tag tag-staff">yes</span>';
}

/* THE FINDING, RATHER THAN THE DATE. A role nobody has used is what somebody
   opens this screen to find, and a column of dates makes them do the
   subtraction — on the row that is easiest to skip past, because nothing about
   it looks wrong.

   NINETY DAYS IS THIS SCREEN'S OWN LINE and it is drawn, not enforced: nothing
   is refused by it and nothing is recorded against it, so it is a reading aid
   rather than a threshold a row was judged by (K-16). A number that decided
   something would belong beside the row it decided about. */
const QUIET_DAYS = 90;

function used(one, gone) {
  if (!one.last_opened_console) {
    return '<span class="tag tag-warn">never</span>' +
      (gone ? '' : '<span class="cell-sub">granted and not used</span>');
  }

  const when = new Date(one.last_opened_console);
  const days = Math.floor((Date.now() - when.getTime()) / 86400000);

  return '<span class="cell-main">' + esc(day(one.last_opened_console)) + '</span>' +
    (!gone && days >= QUIET_DAYS
      ? '<span class="cell-sub">' + days + ' days ago</span>'
      : '');
}

function day(iso) {
  if (!iso) return '—';
  const at = new Date(iso);
  return Number.isNaN(at.getTime()) ? String(iso) : at.toLocaleDateString();
}

File: internal/console/staff.go
package console

import (
	"context"
	"net/http"
	"time"

	"github.com/google/uuid"

	"github.com/codeschool-ing/schooling/internal/platform/web"
)

/* Who can open this console.

   # THE ONE QUESTION THIS CONSOLE COULD NOT ASK ABOUT ITSELF

   `cmd/staff list` has printed this since phase 0 and prints it into a
   terminal, next to the database, on a machine somebody has to be sitting at.
   So the console could set what every student pays, erase a person, quarantine
   a question and read every entry in the audit — and could not answer "who else
   can do all of that". That is not a small gap. An access list nobody can see
   is an access list nobody reviews, and the row that matters on it is always
   the one nobody remembered was there.

   # IT IS A READ AND IT STAYS A READ

   Nothing here grants or revokes. That is not squeamishness about a form: the
   first owner cannot be granted a role by a console that needs a role to open,
   so `cmd/staff` has to exist regardless — and two doors that both write to
   `staff` is two paths to keep audited, tested and in agreement, bought for the
   convenience of not opening a terminal on the rarest write this platform has.
   The screen says where the grant lives instead.

   `writes.go` lists what this console can CHANGE, and this route is correctly
   absent from it.

   # WHY THIS IS NOT THE THING K-22 REFUSES

   K-22 says a person is found by an exact address and never listed, because
   browsing personal data is indistinguishable from working. The argument does
   not reach this list, and the difference is not one of degree:

     - it is not a population, it is this platform's access-control list, and
       there is no version of reviewing one that goes address by address — the
       whole question is who is on it that you did not think to ask about;
     - everybody on it consented to operating the platform rather than to using
       it, and their address is already in every audit entry they have written;
     - it is bounded by the number of people who run this, which is a number
       that fits on one screen and always will.

   # AND IT IS NOT AUDITED, WHICH IS A DECISION AND NOT AN OMISSION

   K-20 audits the personal-data export because afterwards a person's record is
   a file on a laptop, outside every control this system has. Nothing here
   leaves: it is the roster, read by people who are on it, and an entry per
   read would be an audit trail of operators looking at their own names — noise
   in exactly the log that has to stay searchable for the week it matters.

   # WHAT IT SHOWS THAT `cmd/staff list` DOES NOT

   Who granted it, whether the row was revoked, and when they last presented a
   second factor. The last of those is the one an access review is actually for:
   a role nobody has used in a year is the row to ask about, and neither the
   role nor the enrolment says a word about it.
*/

// Operator is one person with a role, as this console shows them.
type Operator struct {
	AccountID uuid.UUID
	Name      string
	Email     string
	Role      string

	GrantedAt time.Time

	// Who granted it, or empty for the first owner — who has nobody above them
	// and whose row says so with a null rather than with a name.
	GrantedByName  string
	GrantedByEmail string

	// Set on somebody who left. Revoked rows are shown rather than filtered,
	// because "who has ever been able to open this" is the question with an
	// audit in it and the one a screen is for.
	RevokedAt *time.Time

	// Whether they have a second factor at all. A role without one opens
	// nothing.
	SecondFactor bool

	// When they last presented it, which is when they last actually opened this
	// console — nil for somebody who never has.
	LastOpenedConsole *time.Time
}

// Operators is what this package may not import: `identity` owns the table.
type Operators func(ctx context.Context) ([]Operator, error)

// StaffHandler answers who can open this console.
//
// IT TAKES NO `Record`, NO `Label` AND NO RANK CLOSURE, and the shortness of
// this struct is the design rather than a stage it has not reached. There is
// nothing to record because nothing is written; there is nobody to label
// because no entry is produced; and there is no second rank because read-only
// is the floor of the console and reviewing who has access is the most
// read-only act there is.
type StaffHandler struct {
	operators Operators
}

func NewStaffHandler(operators Operators) *StaffHandler {
	return &StaffHandler{operators: operators}
}

func (h *StaffHandler) Routes(mux *http.ServeMux) {
	mux.HandleFunc("GET /console/api/v1/staff", h.list)
}

type operatorBody struct {
	ID    string `json:"id"`
	Name  string `json:"name"`
	Email string `json:"email"`
	Role  string `json:"role"`

	GrantedAt      time.Time `json:"granted_at"`
	GrantedByName  string    `json:"granted_by_name,omitempty"`
	GrantedByEmail string    `json:"granted_by_email,omitempty"`

	// Absent rather than null on somebody still here: the screen draws a row
	// differently when there is one, and `omitempty` is that rule said once.
	RevokedAt *time.Time `json:"revoked_at,omitempty"`

	// NOT `omitempty`. False is the answer that matters most on this screen —
	// a role that opens nothing — and a key the interface failed to read must
	// not look like it.
	SecondFactor bool `json:"second_factor"`

	// Absent on somebody who has never opened the console, which the screen
	// says in words. A zero time would be drawn as a date in the year 1.
	LastOpenedConsole *time.Time `json:"last_opened_console,omitempty"`
}

func (h *StaffHandler) list(w http.ResponseWriter, r *http.Request) {
	people, err := h.operators(r.Context())
	if err != nil {
		web.LoggerFrom(r.Context()).Error("reading who can open the console", "error", err)
		web.Fail(w, http.StatusServiceUnavailable, web.CodeInternal, "could not read that")
		return
	}

	out := make([]operatorBody, 0, len(people))
	for _, one := range people {
		out = append(out, operatorBody{
			ID: one.AccountID.String(), Name: one.Name, Email: one.Email, Role: one.Role,
			GrantedAt:      one.GrantedAt,
			GrantedByName:  one.GrantedByName,
			GrantedByEmail: one.GrantedByEmail,
			RevokedAt:      one.RevokedAt,
			SecondFactor:   one.SecondFactor,

			LastOpenedConsole: one.LastOpenedConsole,
		})
	}

	web.JSON(w, http.StatusOK, map[string]any{
		"staff": out,

		/* WHERE THE GRANT LIVES, SENT RATHER THAN WRITTEN INTO THE SCREEN.
		   Somebody reading this list is one thought away from wanting to change
		   it, and a screen that says nothing leaves them looking for a button
		   that is deliberately not there. The command is the answer, and it is
		   here so that the reason travels with it. */
		"how_to_change_it": "A role is granted and revoked with `staff grant` and " +
			"`staff revoke`, from a terminal where the database is reachable. It is not a " +
			"form here because the first owner cannot be granted a role by a console that " +
			"needs one to open — so that door has to exist anyway, and a second door onto " +
			"the same table would be a second thing to keep audited and in agreement.",

		/* AND WHY THE ROSTER IS NOT AN ALARM. `job_runs` has the same sentence
		   for the same reason: a screen answers "should I trust this" for
		   somebody already looking, and it cannot answer "has something
		   changed" for somebody who is not. */
		"about_reviewing": "Every row here is somebody who can open this console, or could. " +
			"The column worth reading is the last one: a role granted a year ago and never " +
			"used since is access nobody is missing, and it is the row an access review " +
			"exists to find. Revoked rows stay so that somebody who left is distinguishable " +
			"from somebody who was never here.",

		// A DATABASE WITH NO STAFF IS UNREACHABLE FROM HERE, which makes this
		// sentence a statement about a defect rather than an empty state. The
		// request that arrived was authenticated by a live staff row, so the
		// reader is on the list they are being told is empty.
		"impossible": "Nobody has a role, which cannot be true — this page was opened by " +
			"somebody who has one. Something is answering wrongly rather than there being " +
			"nobody here.",
	})
}
